pages/web-scraping/index.jsx

```diff
@@ -0,0 +1,249 @@
+  { city: 'London', code: 'LHR' },
+  { city: 'Tokyo', code: 'HND' },
+  { city: 'Paris', code: 'CDG' },
+  { city: 'Hong Kong', code: 'HKG' },
+  { city: 'Singapore', code: 'SIN' },
+  { city: 'Dubai', code: 'DXB' },
+  { city: 'Shanghai', code: 'PVG' },
+  { city: 'Seoul', code: 'ICN' },
+  { city: 'Bangkok', code: 'BKK' },
+  { city: 'Amsterdam', code: 'AMS' },
+  { city: 'Beijing', code: 'PEK' },
+  { city: 'Frankfurt', code: 'FRA' },
+  { city: 'Cape Town', code: 'CPT' },
+  { city: 'Sydney', code: 'SYD' },
+  { city: 'Melbourne', code: 'MEL' },
+  { city: 'Toronto', code: 'YYZ' },
+  { city: 'Vancouver', code: 'YVR' },
+  { city: 'Montreal', code: 'YUL' },
+  { city: 'Brussels', code: 'BRU' },
+  { city: 'Copenhagen', code: 'CPH' },
+  { city: 'Oslo', code: 'OSL' },
+  { city: 'Stockholm', code: 'ARN' },
+  { city: 'Helsinki', code: 'HEL' },
+  { city: 'Rome', code: 'FCO' },
+];
+
+/**
+ * @typedef WebScrapingUseCaseProps
+ * @property {Object} query
+ */
+
+/**
+ * @param {WebScrapingUseCaseProps} props - Props for the component
+ * @returns {JSX.Element} React component
+ */
+export const WebScrapingUseCase = ({ query }) => {
+  const [from, setFrom] = useState(AIRPORTS[0].code);
+  const [to, setTo] = useState(AIRPORTS[1].code);
+
+  /** @typedef {import('../../client/components/web-scraping/FlightCard').Flight} Flight */
+  /** @type {[Flight[] | undefined, React.Dispatch<Flight[] | undefined>]} */
+  const [flights, setFlights] = useState();
+  const [message, setMessage] = useState('');
+
+  /** @typedef {import('../../server/server').Severity} Severity */
+  /** @type {[Severity | undefined, React.Dispatch<Severity>]} */
+  const [messageSeverity, setMessageSeverity] = useState();
+  const [loading, setLoading] = useState(false);
+
+  /**
+   * Use the Fingerprint Pro React SDK hook to get visitor data (https://github.com/fingerprintjs/fingerprintjs-pro-react)
+   * For Vue, Angular, Svelte, and other frameworks, see https://dev.fingerprint.com/docs/frontend-libraries
+   * See '/client/use-visitor-data.js' for an example implementation of similar functionality without the SDK
+   */
+  const { getData } = useVisitorData(
+    {
+      products: query.backdoor ? ['identification'] : ['botd'],
+      // Don't use a cached fingerprint, it must be fresh to avoid replay attacks
+      ignoreCache: true,
+    },
+    // Don't fingerprint the visitor on mount, but when they click "Search flights", the fingerprint must be fresh
+    { immediate: false }
+  );
+
+  /**
+  /* @type {React.FormEventHandler<HTMLFormElement>}
+  /*/
+  async function handleSubmit(event) {
+    event.preventDefault();
+    setLoading(true);
+    setMessage('');
+    const visitorData = await getData();
+    console.log(visitorData);
+    try {
+      /** @type {import('../../server/checkResult').CheckResult} */
+      const result = await (
+        await fetch(`/api/web-scraping/flights?from=${from}&to=${to}&requestId=${visitorData.requestId}`)
```

Comment:
We use `react-query` for API calls, to simplify error and loading state handling. You can check how we do that in `/client/api` directory, it should simplify this code a little bit :)

Reply:
Okay, I made it work with React query, here is the implementation: 

<details>
  <summary>See implementation</summary>
  
  ```js
  const getFlightsQuery = useMutation(
    ({ requestId }) =>
      fetch(`/api/web-scraping/flights`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          from: fromCode,
          to: toCode,
          requestId,
        }),
      }).then((res) => res.json()),
    {
      onSuccess: (result) => {
        setLoading(false);
        setFlights(result.data);
        if (result?.severity !== 'success') {
          setMessage(result.message);
          setMessageSeverity(result.severity);
        }
      },
      onError: (error) => {
        setLoading(false);
        setMessageSeverity('error');
        setMessage(error.toString());
        setFlights([]);
      },
    }
  );

  async function searchFlights() {
    setLoading(true);
    setMessage('');
    const { requestId } = await getVisitorData();
    getFlightsQuery.mutateAsync({ requestId });
  }
  ```
</details>

Is there a better way to do it than I am missing? I only have limited experience with react-query, but in this case I don't think it simplifies things. I still need to manage loading in state because it's triggered when I get fingerprint data, not just the flight data. It's more lines, two functions instead of one and the syntax will be unfamiliar to users who haven't used it before. 

![image](https://user-images.githubusercontent.com/11249993/222430592-a97709db-9547-494f-885f-9586eeb4cd01.png)

Sure, you could hide the implementation details in other files, but my thesis was the user shouldn't have to dig through multiple files and function calls to figure out what's going on when what's going on is as simple as an HTTP request.

Happy to commit the react-query version to keep it consistent with other use cases :) just wondering if I am missing a better way to write it.

Reply:
It's a good start! Although, you should use `useQuery` hook instead of `useMutation`, which will let you get rid of `loading`, `messageSeverity`, `message` and `flights` variables, because you will be able to access those from the query itself :)